devito/at_controller.py

```diff
@@ -0,0 +1,225 @@
+        :param minimum: int - minimum tune range
+        :param maximum: int - maximum tune range
+        """
+        timing_run = 0  # estimating run time
+        for i in range(0, 5):
+            logger.info('Estimating auto-tuning runtime...sample %d/5' % (i + 1))
+
+            blocks = []
+            for block in self.op.propagator.block_sizes:
+                blocks += [random.randrange(minimum, maximum)] if block else [None]
+
+            self.op.propagator.block_sizes = blocks
+            timing_run += self.get_execution_time()
+
+        logger.info("Estimated runtime: %f minutes." % float(
+            timing_run / 5 * ((maximum - minimum) * (maximum - minimum)) / 600))
+
+    def get_execution_time(self):
+        """
+        Runs and times the function
+        :return: time - how long it took to execute
+        """
+        self.op.propagator.run(self.op.get_args())
+        return self.op.propagator.timings['kernel']
+
+    def _write_block_report(self, times):
+        """
+        Writes auto tuning report for block sizes
+        :param times: sorted list - times with block sizes
+        :raises IOError: if fails to write report
+        """
+        global model_descr_template
+        try:
+            full_report_text = ["%s %f\n" % (str(block), timee) for block, timee in times]
+
+            # Cache blocking dimensions
+            cb_dims = str([True if block else False for block in self.op.propagator.block_sizes]).replace(" ", '')
+            shape = str(self.op.shape).replace(" ", '')
+
+            # Writes all auto tuning information into full report
+            with open(path.join(self.report_dir, "%s_time_o_%s_spc_bo_%s_shp_%s_b_%s_report.txt" %
+                      (self.op.getName(), self.op.time_order, self.op.spc_border, shape, cb_dims)), 'w') as f:
+                f.writelines(full_report_text)
+
+            # string that describes the model
+            model_descr_str = model_descr_template % (self.op.getName(), self.op.time_order,
+                                                      self.op.spc_border, shape, cb_dims)
+            str_to_write = "%s %s\n" % (model_descr_str, str(times[0][0]).replace(" ", ''))
+
+            if not path.isfile(self.final_report_path):  # initialises report file if it does not exist
+                with open(self.final_report_path, 'w') as final_report:
+                    final_report.write("f name, time o, space bo ,shape, block dims, best block size\n")
+                    final_report.write(str_to_write)  # writes the string
+                return
+            else:
+                with open(self.final_report_path, 'r') as final_report:  # reads all the contents
+                    lines = final_report.readlines()
+
+                # checks whether entry already exist and updates it. Otherwise appends to the end of the file
+                entry_found = False
+
+                for i in range(1, len(lines)):
```

Comment:
Have you considered using a non-custom file format to utilise external packages, like `json` or numpy's `csv` reader? I personally like `json` which effectively pretty-prints and reads nested Python dictionaries, which can be made to be nicely human readable and the read in result data can be directly accessed from the resulting dict. For example you could do:
```
profile = {'timestep': <my time stamp>,
           'block_size_range': ...
           'data' = {(3, 6. None): <timing>,
                     (6, None, 7): <timing>,
                      ...
                     }
           }